### PR TITLE
Configuring flashgg for STXS 1.1

### PR DESCRIPTION
**Pull request by jonathon-langford**
> New Rivet tool for 1.1 processes integrated into CMSSW_10_5_0: (see https://github.com/jonathon-langford/cmssw/blob/cmssw_STXS_1p1/SimDataFormats/HTXS/interface/HiggsTemplateCrossSections.h). This is required upstream from flashgg.

> New HiggsClassification object automatically enters microAOD file when produced.

> Updated TagTruthBase and all taggers to accommodate new HiggsClassification object 

> Configured ntuple dumper (Taggers/test/vbf_dumper_2017.py_ to work with with new metaConditions option. Example command:
cmsRun vbf_dumper_2017.py metaConditions=$CMSSW_BASE/src/flashgg/MetaData/data/MetaConditions/Era2016_RR-17Jul2018_v1.json maxEvents=-1

### CODE REVIEW

**Comment by simonepigazzini** [comment]
Hi @jonathon-langford,

thank you for this. Does this include also updates from Ed about the recently approved STXS results?
Is this expect to impact anything else than STXS analyis? At a first look it doesn't sees so, which is good.

Few technical points:
- please add the Emacs and Vi snippets at the bottom of any newly added c++ file to ensure that the 
  indentation is applied correctly by the automatic hook (please also check that the hook is correctly setup 
  in your working area, i.e. did you run the flashgg setup script?)
- rollback the test file names to file:myMicroAOD.root instead of your custom ones.

File: Taggers/test/diphotonsDumper_cfg.py [hunk 1]
```diff
@@ -9,7 +9,7 @@
 process.load("FWCore.MessageService.MessageLogger_cfi")
 process.source = cms.Source("PoolSource",
                             fileNames=cms.untracked.vstring(
-        "file:myMicroAODOutputFile.root"        
+        "file:/afs/cern.ch/work/j/jlangfor/hgg/stxs/1p1/CMSSW_10_2_9/src/flashgg/myMicroAODOutputFile.root"
```
Comment (simonepigazzini) [inline Taggers/test/diphotonsDumper_cfg.py]: rollback

File: Taggers/test/vbf_dumper_2017.py [hunk 2]
```diff
@@ -187,15 +170,7 @@
...
-                                 #"/store/group/phys_higgs/cmshgg/sethzenz/flashgg/RunIISummer16-2_4_1-25ns_Moriond17/2_4_1/VBFHToGG_M125_13TeV_amcatnlo_pythia8_CUETP8M1Down/RunIISummer16-2_4_1-25ns_Moriond17-2_4_1-v0-RunIISummer16MiniAODv2-PUMoriond17_80X_mcRun2_asymptotic_2016_TrancheIV_v6_ext1-v1/170114_093048/0000/myMicroAODOutputFile_2.root"
-#                                 "/store/group/phys_higgs/cmshgg/sethzenz/flashgg/RunIISummer16-2_4_1-25ns_Moriond17/2_4_1/VBFHToGG_M125_13TeV_amcatnlo_pythia8/RunIISummer16-2_4_1-25ns_Moriond17-2_4_1-v0-RunIISummer16MiniAODv2-PUMoriond17_80X_mcRun2_asymptotic_2016_TrancheIV_v6_ext2-v1/170114_092925/0000/myMicroAODOutputFile_19.root"
-
-                                 #"file:/vols/cms/jwright/ForYacine/myMicroAODOutputFile_1.root"
-                                 #"/store/group/phys_higgs/cmshgg/ferriff/flashgg/RunIISpring16DR80X-2_3_0-25ns_Moriond17_MiniAODv2_p3/2_3_0/DoubleEG/RunIISpring16DR80X-2_3_0-25ns_Moriond17_MiniAODv2_p3-2_3_0-v0-Run2016G-23Sep2016-v1/170116_174903/0000/myMicroAODOutputFile_4.root"
+"file:/afs/cern.ch/work/j/jlangfor/public/hgg/stxs/1p1/test_microAOD/vbf_2016_microAOD.root"
```
Comment (simonepigazzini) [inline Taggers/test/vbf_dumper_2017.py:173]: it's fine as a placeholder but, to be updated once a catalogue with the new STXS info will be available

**Comment by jonathon-langford** [comment]
Hi Simone,


No problem! Was just composing an email to explain things further but you beat me to it :)


Firstly, no this does not include the updates from the HIG-18-029 analysis. Ed plans to make a PR after he has finished writing his thesis.


I was discussing this with Ed a few hours ago (so will cc him in), but this may indeed impact the recent STXS analysis. I have integrated the updated Rivet tool [1] upstream in CMSSW, which replaces the stage 1 information with 1.1. You can see the changes here: [2]. This means that the updates from Ed in their current state will not work after this PR, however he said he could make changes to his stuff to accommodate the 1.1. What do you think about this?


Concerning the upstream changes: it may not be appropriate to make an offical PR to CMSSW right now, since other analyses are still using the old stage 1 process definitions. Can you instead add a line to setup_flashgg.sh to merge these changes? You may need to add the following packages as well:

    > git cms-addpkg SimDataFormats/HTXS
    > git cms-addpkg GeneratorInterface/RivetInterface


For the technical points:


1) Are you referring to the newly added Rivet code: HiggsTemplateCrossSections.cc in CMSSW. Apart from that, there are no new c++ files in flashgg. And, yes I ran the flashgg setup script.


2) Will do. I don't think the diphotonsDumper_cfg.py file is configured for the new metaConditions anyway. Only the vbf_dumper_2017.py.


Cheers,

Jonathon

[1] - https://gitlab.cern.ch/LHCHIGGSXS/LHCHXSWG2/STXS
[2] - https://github.com/jonathon-langford/cmssw/commit/23a4171fb563040559919c6c3625b5738aa38bde
________________________________

File: Taggers/test/vbf_dumper_2017.py (hunk 2 above)
Comment (jonathon-langford) [reply Taggers/test/vbf_dumper_2017.py:173]: Yep, agreed.

File: Taggers/test/diphotonsDumper_cfg.py (hunk 1 above)
Comment (jonathon-langford) [reply Taggers/test/diphotonsDumper_cfg.py]: done

**Comment by simonepigazzini** [comment]
Hi Jonathon,

thanks. I won't merge this PR right away, I will wait after Ed is done with the thesis writing to understand how we want to move forward technically. It's fine to do a merge topic to update Rivet.

Concerning the issue with indentation I probably got confused with another PR since I cannot find it anymore :)

simone

**Comment by jonathon-langford** [comment]
Hi Simone,


Agreed, no rush to get this in at the moment so better to wait.


Cheers,

Jonathon

________________________________

**Comment by simonepigazzini** [comment]
@edjtscott @jonathon-langford , I think I am going to merge this before running the upcoming production. I tested the merge-topic to get the needed modification in cmssw and it works fine, I will fix the conflict and test the PR. Do you have any objection? @edjtscott is there anything else from STXS that is needed for MicroAOD production?

**Comment by jonathon-langford** [comment]
@simonepigazzini there has been an update upstream so the new HiggsClassification object in CMSSW includes both stage 1 and stage 1.1 info (https://github.com/cms-sw/cmssw/pull/26724). So you no longer need to make the "upstream" changes I suggested above. The new object should come automatically in the uAOD. 

Not had time to check everything is compatible with the changes I made to the Taggers (i.e. the naming scheme of the functions: e.g. stage1_1_cat(). Also, since the new HiggsClassification also includes stage1 then the current flashgg should run without errors. Therefore I would suggest, not merging this PR just yet

**Comment by jonathon-langford** [comment]
I would try doing a test run of the uAOD production using the lastest CMSSW (or merging https://github.com/cms-sw/cmssw/pull/26724). If the HiggsClassification object contains all stage 1 and 1.1, then I would go ahead with the production.

**Comment by simonepigazzini** [comment]
Thanks for the feedback. Yacine PR has been merged but we will wait for the next stable release before switching.

**Comment by simonepigazzini** [comment]
@jonathon-langford I lost track of this PR, shall I merge this one or are you going to make a new one once the "parent dset" workflow is tested?

**Comment by jonathon-langford** [comment]
@simonepigazzini, me and @edjtscott will make a new PR after testing the "parent dset" workflow, which will also integrate the STXS code that Ed used for the HIG-18-029 analysis. 

Also, this current PR only accommodates stage 0 and 1.1 in the truth objects. For the new one we will have all stages (0, 1 and 1.1).

When we make the new PR, you can close this one.

**Comment by simonepigazzini** [comment]
Closing, STXS will be integrated in a future PR.

**Pull request closed without merging**